Add MultiDeviceTestContext for testing more than one Device

reszelaz opened the pull request:
First of all thanks to @ajoubertza and @tiagocoutinho for your suggestions to the interface.  Also, if you find that the proposed interface could be improved just let me know and I will change it.

DeviceTestContext runs just one Device hence is only useful for unit testing of one Device. Integration tests involving multiple Devices within the same server are not possible. Add MultiDeviceTestContext for that purpose. This context thanks to a little bit more complicated interface allows to run multiple Devices (each of them even exporting multiple devices) all within one server.

At the same time refactor the DeviceTestContext to use the MultipleDeviceTestContext as its base class. This way, the DeviceTestContext is just a simplified interface to its base class.

reszelaz committed:
Add MultiDeviceTestContext for testing more than one Device

DeviceTestContext runs just one Device hence is only useful for unit testing
of one Device. Integration tests involving multiple Devices within the same
server are not possible. Add MultiDeviceTestContext for that purpose. This
interface thanks to a little bit more complicated interface allows to run
multiple Devices (each of them even exporting multiple devices) all within
one server.

At the same time refactor the DeviceTestContext to use the
MultipleDeviceTestContext as its base class. This way, the DevieTestContext
maintains is just a simplified interface to its base class.

reszelaz committed:
Fix super call in python 2

reszelaz commented:
An example of usage:

```python
from tango import DeviceProxy
from tango.test_context import MultiDeviceTestContext

from sardana.tango.pool.Pool import Pool, PoolClass
from sardana.tango.pool.Controller import Controller, ControllerClass


devices_info = (
    {
        "class": (PoolClass, Pool),
        "devices": (
            {
                "name": "pool/test/1"
            },
        )
    },
    {
        "class": (ControllerClass, Controller),
        "devices": (
            {
                "name": "controller/test/1",
                "properties": {
                   "Id": 1,
                   "Klass": "DummyMotorController",
                   "Library": "DummyMotorController.py",
                   "Type": "Motor"
                }
            },
        )
    }
)

def test_controller():
    """Test device voltage reading after calibration."""
    with MultiDeviceTestContext(devices_info, process=True, host="localhost") as context:
        pool = context.get_device("pool/test/1")
        print(pool.ControllerList)
```

ajoubertza commented on tango/test_context.py:
```diff
@@ -95,37 +96,49 @@ def get_host_ip():
+                  the following keys:
+                  * "name" (str)
+                  * "properties" (dict)
+        """
```
Interface looks good.

ajoubertza commented on tango/test_context.py:
```diff
@@ -136,27 +149,46 @@ def __init__(self, device, device_cls=None, server_name=None,
+            if device_cls:
+                class_list.append((device_cls, device, tangoclass))
+            else:
+                device_list.append(device)
```
Looking at this, shouldn't we move `device_cls  = None`  (L163) inside the loop?  We might have input like this, where later devices don't specify a class:
```python
devices_info = (
    {
        "class": (PoolClass, Pool),
        "devices": (
            {
                "name": "pool/test/1"
            },
        )
    },
    {
        "class": MyDevice,
        "devices": (
            {
                "name": "my/device/1"
            },
        )
    }
)
```

reszelaz replied:
You are right! I'm correcting this. Thanks for spotting that!

reszelaz replied:
Done in b3c6316.
But actually this context is not yet ready for mixing both: HLAPI and classical API.
If someone wants in the future to enhance it this part will be already done:)

ajoubertza commented on tango/test_context.py:
```diff
@@ -136,27 +149,46 @@ def __init__(self, device, device_cls=None, server_name=None,
+            tangoclass = device.__name__
+            # File
+            self.generate_db_file_tangoclass(server_name, instance_name,
+                                             tangoclass)
```
`server_name` and `instance_name` don't change during the loop, while `tangoclass` will.  I'm guessing it will be a problem if someone specifies a config that results in the same `tangoclass` more than once.  Will it? If so, should we do some validation, or will the problem be obvious from the failure?

Alternatively, could we make `tangoclass` unique, or must it be `device.__name__`?

reszelaz replied:
Yes, configuration resulting in repeating the tangoclass fail as it is now. Furthermore the error message is not intuitive at all. I will add some validation.

reszelaz replied:
Done in: f6286e3

ajoubertza replied:
OK.   Thanks for including the commit hash - very easy to review

ajoubertza commented on tango/test_context.py:
```diff
@@ -136,27 +149,46 @@ def __init__(self, device, device_cls=None, server_name=None,
+            runserver = partial(run, device_list, cmd_args)
+        else:
+            raise ValueError("Wrong format of devices_info")
+
```
Maybe we should warn or fail if both `class_list` and `device_list` are not empty.  In that case, some of the devices requested by the user will not be started up.  It looks like that combination is not supported.

reszelaz replied:
You are right! I'm adding this validation. IMO warn is too soft, I would raise a `ValueError`.

reszelaz replied:
Done in: e693575

ajoubertza commented on tango/test_context.py:
```diff
@@ -197,29 +229,63 @@ def generate_db_file(self, server, instance, device,
         """Generate a database file corresponding to the given arguments."""
```
I think we can remove this method.  It isn't used anymore.  I can't see why someone would want to use it externally to the `DeviceTestContext` initialisation either.

reszelaz replied:
Ok, I left it just in case only for backwards compat. I'm removing it.

ajoubertza commented on tango/test_context.py:
```diff
@@ -197,29 +229,63 @@ def generate_db_file(self, server, instance, device,
+        with open(self.db, "a") as f:
             f.write("/".join((server, instance, "DEVICE", tangoclass)))
-            f.write(': "' + device + '"\n')
+            f.flush()
```
Is the `flush` necessary, given that exiting the `open` context handler closes the file?  Same for the next method.

reszelaz replied:
No, I don't remember now how it got there:) I just tested my examples without and it sill works.
I just removed it in 0ade4703.

ajoubertza commented on tango/test_context.py:
```diff
@@ -197,29 +229,63 @@ def generate_db_file(self, server, instance, device,
+        device_names = [info["name"] for info in device_prop_info]
+        with open(self.db, "a") as f:
+            for device_name in device_names:
+                f.write(': "' + device_name + '"\n')
```
If there are multiple devices we'd end up with a file like this:
```
Clock/clock/DEVICE/Clock: "test/nodb/clock1"
: "test/nodb/clock2"
```
Will that work?  I'm not sure about the format of these files.

ajoubertza replied:
I'm still wondering about this?

ajoubertza replied:
> I don't know how to reply to this: #314 (comment) so I generate a new comment here...
>
> What do you exactly mean? Whether the MultiDeviceTestContext.append_db_file method name is fine? or maybe its API? Or you reconsider having two methods as it was before #314 (comment).

I'm thinking about the code that will parse this file.  My question is about the format of the line with `DEVICE` in it.  The current version of your `append_db_file` method will have the first device name on the same line as the `DEVICE` part.  Then the next device is on a new line:
```
Clock/clock/DEVICE/Clock: "test/nodb/clock1"
: "test/nodb/clock2"
```

It is just a bit inconsistent, so I was wondering if it works.  Something like this would be more "consistent" for me:
```
Clock/clock/DEVICE/Clock
: "test/nodb/clock1"
: "test/nodb/clock2"
```

I tried exporting some properties from Jive to see the file format, but the files didn't have a `DEVICE` line.

reszelaz replied:
Sorry I didn't see your initial comment. Now I understand your concern.
Actually now testing it with two devices of the same class I have realized that it does now work correctly. I will investigate it tomorrow.

ajoubertza replied:
The update in 7c7a649 for multiple devices look good.  Thanks.

ajoubertza commented on tango/test_context.py:
```diff
@@ -197,29 +229,63 @@ def generate_db_file(self, server, instance, device,
+        return form.format(self.host, self.port, device_name, self.nodb)
+
+    def get_device(self, device_name):
+        """Return the device proxy corresponding to the given device name."""
```
Suggest:  `"""Return new device proxy ..."""`
Important to know that a new instance is being instantiated by calling this method.

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +333,75 @@ def __enter__(self):
         """Exit method for context support."""
         self.stop()
 
+# Device test context
```
Suggest:   `# Single device test context`

reszelaz replied:
Ok

ajoubertza commented on tango/test_context.py:
```diff
@@ -197,29 +229,63 @@ def generate_db_file(self, server, instance, device,
+        )
+        return self.generate_db_file_device(device_prop_info)
+
+    def generate_db_file_tangoclass(self, server, instance, tangoclass):
```
If we remove the old `generate_db_file` method, then I think we might as well combine this method with `generate_db_file_device`.   I don't like that `generate_db_file_tangoclass` leaves the file in an incomplete state.

New combined method can then be called `generate_db_file` or maybe `update_db_file` or `append_db_file`.

reszelaz replied:
Ok, I merge these two methods `generate_db_file_tangoclass` and `generate_db_file_device` into one:
```python
def append_db_file(self, server, instance, tangoclass, device_prop_info):
    [...]
```

reszelaz replied:
Done in:  0764edb

ajoubertza commented on tango/test_context.py:
```diff
@@ -250,9 +316,6 @@ def connect(self):
         self.server.ping()
-        # Get device proxy
-        self.device = DeviceProxy(self.get_device_access())
-        self.device.ping()
```
I wonder if not waiting to ping all the devices might cause some problems when using this for unit tests.  Maybe the server is up but not all the devices.  What about making a dict of `DeviceProxy` instances for all the devices, and pinging them here?   Then we wouldn't need to make a new proxy on each call to `get_device()`.

reszelaz replied:
Sounds good. I'm proceeding with the change.

reszelaz replied:
Actually, with the tests that I did, I observe that the ping to dserver device replies only after all the other devices are exported. Another note is that the ping is not "waiting". If there is no connection, the ping simply raises an exception. 

IMO, ping to all the others devices is "just in case". Do you know any example of what could go wrong so the other devices does not respond to ping and the dserver responds?

I imagine that in Sardana we will make a generic test context that will populate "sar_demo" device server with approx. 30 devices. Then in particular tests we will just use proxies to, let's say, a couple of them. I think that preparing all the 30 device proxies up-front is not optimal. What I see good is to maintain cache of the device proxies within the context so it will be filled gradually on calls to `get_device`. Or simple don't maintain this cache and leave it to the test developer to keep them and not call `get_device` every time it is necessary to do something with the device.

I'll wait with implementing these changes for your opinion.

ajoubertza replied:
> Actually, with the tests that I did, I observe that the ping to dserver device replies only after all the other devices are exported. Another note is that the ping is not "waiting". If there is no connection, the ping simply raises an exception.

OK, good to know.  OK, I see, the ping will not wait.

> IMO, ping to all the others devices is "just in case". Do you know any example of what could go wrong so the other devices does not respond to ping and the dserver responds?

No, no examples.

> I imagine that in Sardana we will make a generic test context that will populate "sar_demo" device server with approx. 30 devices. Then in particular tests we will just use proxies to, let's say, a couple of them. I think that preparing all the 30 device proxies up-front is not optimal. What I see good is to maintain cache of the device proxies within the context so it will be filled gradually on calls to `get_device`. Or simple don't maintain this cache and leave it to the test developer to keep them and not call `get_device` every time it is necessary to do something with the device. 

That's an interesting use case.  With a little effort, we could get the best of both.  I'm thinking we could create a dict-like object that lazy loads the device proxies.  We return this to the user (or they access it via a `self.devices` property).  The first time a key is accessed, the `DeviceProxy` object is created.  What do you think?

reszelaz replied:
Done in 0dac6a9.

Finally, to make it simpler, I went for an option for directly keeping the dictionary within the context class scope. A dedicated class for this purpose would need to know the host, port, nodb to be able to compose the Tango URI. This is because we want to get the devices using just the family/domain/member names, as these are used in the `devices_info` argument.

ajoubertza replied:
OK.

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +333,75 @@ def __enter__(self):
         if not self.thread.is_alive():
             self.start()
-        return self.device
+        return self
```
This is a little different to the `DeviceTestContext`.  We could return the dict of device proxies here.  Although, sometimes it is useful to have access to the whole object (e.g. to access the server, or get the device access URL).  I don't mind, either way.

reszelaz replied:
IMHO self is enough. Anyway it is somehow related to the previous comment (whether we create device proxies upfront).

ajoubertza replied:
OK, we'll leave it returning the whole object.

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +333,75 @@ def __enter__(self):
+    def connect(self):
+        super(DeviceTestContext, self).connect()
+        # Get device proxy
+        print(self.get_device_access())
```
Think this can be removed (maybe left over from debugging?)

reszelaz replied:
Sure! I'm removing this.

ajoubertza reviewed (changes requested):
Very nice, @reszelaz!  Thanks for the improvement.

reszelaz committed:
Reset device_cls in each iteration

This is just in case someone mixes HLAPI with classical API.

reszelaz committed:
Prevent multiple entries for the same Tango class in devices_info

reszelaz committed:
Merge db file generation methods into one: append_db_file

Also remove generate_db_file method (not used anymore).

reszelaz committed:
Prevent mixing HLAPI and classical API

There are no restrictions in PyTango to mix them but the
MultiDeviceTestContext is not ready yet for doing so.

reszelaz committed:
Correct comment

reszelaz committed:
Remove print

reszelaz committed:
Maintain cache of DeviceProxies for future reuse

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +323,74 @@ def __enter__(self):
+    def connect(self):
+        super(DeviceTestContext, self).connect()
+        # Get device proxy
+        self.device = DeviceProxy(self.get_device_access())
```
We should make use of `get_device` here so that the proxy object is in the cache.  That has the added benefit that accessing the proxy via `self.device` or `self.get_device` returns the same instance.

reszelaz replied:
I agree! Done in 6502e9a4.

Related to this and also to the Codacy error I have done a simple modification of the `DeviceTestContext.get_device_access` API in 9d21b03f. We could also think of deprecating this method and providing `DeviceTestContext.get_my_device_access`, but I don't know if it's worth that?

ajoubertza replied:
I'm happy with your workaround for `get_device_access` as is.

ajoubertza commented on tango/test_context.py:
```diff
@@ -95,37 +96,49 @@ def get_host_ip():
-        """Inititalize the context to run a given device."""
-        # Argument
-        tangoclass = device.__name__
+        """Initialize the context to run given devices within one server.
```
I think we should remove this docstring, and put all the info in the class docstring.  That is what has been done in other parts of the code.  E.g. https://github.com/tango-controls/pytango/blob/492c8752650744b88112bebc450f8497a8b0315b/tango/server.py#L1183-L1204

ajoubertza commented on tango/test_context.py:
```diff
@@ -95,37 +96,49 @@ def get_host_ip():
+    The difference with respect to
+    :class:`~tango.test_context.DeviceTestContext` is that it allows
+    to export multiple devices (even of different Tango classes).
+    """
```
I think we should add some more detail here.  The example is sardana, but maybe that should be changed to some trivial example, so users don't have to install sardana to try this out?

```suggestion
    """
    Example usage::

        from tango.test_context import MultiDeviceTestContext

        from sardana.tango.pool.Pool import Pool, PoolClass
        from sardana.tango.pool.Controller import Controller, ControllerClass

        devices_info = (
            {
                "class": (PoolClass, Pool),
                "devices": (
                    {
                        "name": "pool/test/1"
                    },
                )
            },
            {
                "class": (ControllerClass, Controller),
                "devices": (
                    {
                        "name": "controller/test/1",
                        "properties": {
                           "Id": 1,
                           "Klass": "DummyMotorController",
                           "Library": "DummyMotorController.py",
                           "Type": "Motor"
                        }
                    },
                    {
                        "name": "controller/test/2",
                        "properties": {
                           "Id": 2,
                           "Klass": "DummyMotorController",
                           "Library": "DummyMotorController.py",
                           "Type": "Motor"
                        }
                    },
                )
            }
        )

        def test_controller():
            with MultiDeviceTestContext(devices_info, process=True) as context:
                pool = context.get_device("pool/test/1")
                print(pool.ControllerList)

    :param devices_info:
      a sequence of dicts with information about
      devices to be exported. Each dict consists of the following keys:
        * "class" which value is either of:
          * :class:`~tango.server.Device`
          * a sequence of two elements :class:`~tango.DeviceClass`
            and :class:`~tango.DeviceImpl`
        * "devices" which value is a sequence of dicts with the following keys:
          * "name" (str)
          * "properties" (dict)
    :type devices_info:
      sequence<dict>
    :param server_name:
      Name to use for the device server.
      Optional.  Default is the first device's class name.
    :type server_name:
      :py:obj:`str`
    :param instance_name:
      Name to use for the device server instance.
      Optional.  Default is lower-case version of the server name.
    :type instance_name:
      :py:obj:`str`
    :param db
      Path to a pre-populated text file to use for the
      database.
      Optional.  Default is to create a new temporary file and populate it
      based on the devices and properties supplied in `devices_info`.
    :type db:
      :py:obj:`str`
    :param host:
      Hostname to use for device server's ORB endpoint.
      Optional.  Default is a local IP address.
    :type host:
      :py:obj:`str`
    :param port:
      Port number to use for the device server's ORB endpoint. 
      Optional.  Default is chosen by omniORB.
    :type port:
      :py:obj:`int`
    :param debug:
      Debug level for the device server logging.
      0=OFF, 1=FATAL, 2=ERROR, 3=WARN, 4=INFO, 5=DEBUG.
      Optional. Default is warn.
    :type debug:
      :py:obj:`int`
    :param process:
      True if the device server should be launched in a new process, otherwise
      use a new thread.  Note:  if the context will be used mutiple times, it
      may seg fault if the thread mode is chosen.
      Optional.  Default is thread.
    :type process:
      :py:obj:`bool`
    :param daemon:
      True if the new thread/process must be created in daemon mode.  
      Optional.  Default is not daemon.
    :type daemon:
      :py:obj:`bool`
    :param timeout:
      How long to wait (seconds) for the device server to start up, and also
      how long to wait on joining the thread/process when stopping. 
      Optional.  Default differs for thread and process modes.
    :type timeout:
      :py:obj:`float`
    """
```

reszelaz replied:
Done in 59906332. I didn't know where to put the Device classes implementations so I placed them inline in the example. Also to make it simple I just used HLAPI and didn't use the properties.

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +323,74 @@ def __enter__(self):
         """Enter method for context support."""
```
```suggestion
        """Enter method for context support.

        :return:
          Instance of this test context.  Use `get_device` to get proxy
          access to any of the devices started by this context.
        :rtype:
          :class:`~tango.test_context.MultiDeviceTestContext`
        """
```

reszelaz replied:
Done in aa5ffa20

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +323,74 @@ def __enter__(self):
+# Single device test context
+
+class DeviceTestContext(MultiDeviceTestContext):
+    """ Context to run a device without a database."""
```
```suggestion
    """Context to run a single device without a database.

    The difference with respect to
    :class:`~tango.test_context.MultiDeviceTestContext` is that it only
    allows to export a single device.

    Example usage::
        
        from time import sleep

        from tango.server import Device, attribute, command
        from tango.test_context import DeviceTestContext

        class PowerSupply(Device):

            @attribute(dtype=float)
            def voltage(self):
                return 1.23

            @command
            def calibrate(self):
                sleep(0.1)

        def test_calibrate():
            '''Test device calibration and voltage reading.'''
            with DeviceTestContext(PowerSupply, process=True) as proxy:
                proxy.calibrate()
                assert proxy.voltage == 1.23
      
    :param device:
      Device class to be run.
    :type device:
      :class:`~tango.server.Device` or :class:`~tango.DeviceImpl`
    :param device_cls:
      The device class can be provided if using the low-level API.
      Optional.  Not required for high-level API devices, of type
      :class:`~tango.server.Device`.
    :type device_cls:
      :class:`~tango.DeviceClass`

     The rest of the parameters are described in
     :class:`~tango.test_context.MultiDeviceTestContext`.
    """
```

reszelaz replied:
Done in f558aac1.

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +323,74 @@ def __enter__(self):
+                 instance_name=None, device_name=None, properties=None,
+                 db=None, host=None, port=0, debug=3,
+                 process=False, daemon=False, timeout=None):
+        """Inititalize the context to run a given device."""
```
As above, suggest this is removed, and class-level docstring is used instead.

reszelaz replied:
Done in 43ce4798.

ajoubertza commented on tango/test_context.py:
```diff
@@ -270,12 +323,74 @@ def __enter__(self):
+        self.device.ping()
+
+    def __enter__(self):
+        """Enter method for context support."""
```
```suggestion
        """Enter method for context support.

        :return:
          A device proxy to the device started by this context.
        :rtype:
          :class:`~tango.DeviceProxy`
        """
```

reszelaz replied:
Done in 894b7122

ajoubertza commented on tango/test_context.py:
```diff
@@ -26,9 +26,10 @@
 from . import DeviceProxy, Database, Util
 
-__all__ = ("DeviceTestContext", "run_device_test_context")
+__all__ = ("MultiDeviceTestContext", "DeviceTestContext", "run_device_test_context")
```
I just noticed that this is also done in the test_utils module, so please add something similar there:
https://github.com/tango-controls/pytango/blob/492c8752650744b88112bebc450f8497a8b0315b/tango/test_utils.py#L27

reszelaz replied:
Done in db5de328

reszelaz committed:
Fill cache with the device proxy when connecting

get_device is the proper way to access to devices since this method
fills and reuses the cache to not create and destroy DeviceProxies at
every request.

reszelaz committed:
Add extra argument to DeviceTestContext.get_device_access

DeviceTestContext.get_device_access interfere with the
MultiDeviceTestContext.get_device_access (base class) method. They both should
have a common API. Even if it does not have too much sense to pass a device_name in
DeviceTestContext.get_device_access do it to avoid problems.

reszelaz committed:
Move MultiDeviceTestContext documentation to class docstring

Move MultiDeviceTestContext documentation to class docstring and add
example of usage.

reszelaz committed:
Document MultiDeviceTestContext.__enter__ method.

reszelaz committed:
Document DeviceTestContext class.

reszelaz committed:
Remove DeviceTestContext.__init__ docstring - see class docstring.

reszelaz committed:
Document DeviceTestContext.__enter__.

reszelaz committed:
Make MultiDeviceTestContext importable from tango.test_utils

reszelaz commented:
I don't know how to reply to this: https://github.com/tango-controls/pytango/pull/314#discussion_r339224160 so I generate a new comment here...

What do you exactly mean? Whether the `MultiDeviceTestContext.append_db_file` method name is fine? or maybe its API? Or you reconsider having two methods as it was before https://github.com/tango-controls/pytango/pull/314#discussion_r336793696.

ajoubertza commented on tango/test_context.py:
```diff
@@ -102,6 +102,118 @@ class MultiDeviceTestContext(object):
+                device1 = context.get_device("test/device1/1")
+                device2 = context.get_device("test/device2/1")
+                assert device1.read_attribute("attr1").value == 1.0
+                assert device2.read_attribute("attr1").value == 2.0
```
Nice example, thanks.

ajoubertza commented on tango/test_context.py:
```diff
@@ -192,34 +332,49 @@ def post_init(self):
-            f.write(': "' + device + '"\n')
+            for device_name in device_names:
+                f.write(': "' + device_name + '"\n')
+            f.flush()
```
Is there a need for this flush?  I'd expect exiting the context handler to close the file which will flush any pending writes.

ajoubertza commented on tango/test_context.py:
```diff
@@ -192,34 +332,49 @@ def post_init(self):
+        Maintains previously accessed device proxies in a cache to not recreate
+        then on every access.
+        """
+        if not device_name in self._devices:
```
Minor point, but it will be slightly easier to read as:
```python
if device_name not in self._devices:
```

reszelaz replied:
Done in d30daad5.

ajoubertza reviewed (commented):
This is looking good.  Just my question about the format of the db file that we need to resolve.

ajoubertza commented:
> I don't know how to reply to this: [#314 (comment)](https://github.com/tango-controls/pytango/pull/314#discussion_r339224160) so I generate a new comment here...

I'm not sure why, but I only see the reply option for this on the "Files changed" tab, not this "Conversation" tab.

reszelaz committed:
Remove redundant flush to a file

reszelaz committed:
Fix case with multiple devices of the same class

Multiple devices of the same class must be written in the database file
in the same line, comma-separated. Fix it.

reszelaz commented:
Sorry, I have troubles with replying to your code-review comments. So I reply here for the one about writing in the DB file multiple devices of the same class. I have fixed this in 7c7a6494. In summary, these devices must be written in one line, comma-separated. Now this case works:

```
Pool/pool/DEVICE/Pool: "pool/test/1"
Pool/pool/DEVICE/Controller: "controller/test/1",\
                             "controller/test/2"

#############################################
# CLASS Pool


# CLASS Pool attribute properties


#############################################
# CLASS Controller


# CLASS Controller attribute properties



# DEVICE pool/test/1 properties 


# DEVICE pool/test/1 attribute properties

# DEVICE controller/test/1 properties 

controller/test/1->Id: 1
controller/test/1->Klass: DummyMotorController
controller/test/1->Library: DummyMotorController.py
controller/test/1->Type: Motor

# DEVICE controller/test/1 attribute properties

# DEVICE controller/test/2 properties 

controller/test/2->Id: 2
controller/test/2->Klass: DummyCounterTimerController
controller/test/2->Library: DummyCounterTimerController.py
controller/test/2->Type: CTExpChannel

# DEVICE controller/test/2 attribute properties
```

reszelaz committed:
Change conditional expression to be more understandable

ajoubertza reviewed (dismissed):
Looks good to me!
(Not sure if we can do anything about Codacy's "MultiDeviceTestContext.__init__ is too complex (16)" warning).

I would still like some unit tests for this, but I guess we can add them in a new PR.

reszelaz commented:
> (Not sure if we can do anything about Codacy's "MultiDeviceTestContext.**init** is too complex (16)" warning).

Strange that it started to fail with a commit which has nothing to do with this `__init__`: 0ade4703.

> I would still like some unit tests for this, but I guess we can add them in a new PR.

I can try to add tests. Where should I place them? in `tests/test_context.py` or in `tests/test_test_context.py`?
Is it fine to just use the context to start DS with some _well known_ device(s) and check if the device is correctly exported and behaves as expected? maybe also try to use properties?

ajoubertza commented:
> I can try to add tests. Where should I place them? in `tests/test_context.py` or in `tests/test_test_context.py`?

`tests/test_test_context.py` would follow the pattern we are using: 
`tests/test_ + <thing under test>`

> Is it fine to just use the context to start DS with some _well known_ device(s) and check if the device is correctly exported and behaves as expected? maybe also try to use properties?

You can define a few simple devices in that file.  You will see many examples of `TestDevice` classes created as part of each test in in `test_server.py`.  You'll also see the `server_green_mode` fixture which tries all the the different types, you can include that, if you like.

Yes, just test simple behaviour:
- test that some invalid configurations fail
- test multiple devices (the derived DeviceTestContext is already testing single devices) 
- test get_device_access method
- test get_device and that cache works
- maybe set the State in init_device and then check that it is correct
- test a command
- test reading an attribute
- test that property specified at startup works

ajoubertza committed:
Add unit tests for MultiDeviceTestContext

ajoubertza commented:
@reszelaz I added some tests.  I'll merge in the next day or so, unless you have a comment.

reszelaz commented:
Thanks Anton for writing the test. I am a little bit busy this days. Of course go ahead with the merge!
